### PR TITLE
Make copy if need to pass address to C layer

### PR DESCRIPTION
**Pull request by fufranci**
For memory that will be passed into the C layer, make a copy and store within the C++ secure tunnel object.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: secure_tunneling/CMakeLists.txt [hunk 1]
```diff
@@ -94,7 +94,6 @@ if (NOT IS_SUBDIRECTORY_INCLUDE)
     aws_use_package(aws-crt-cpp)
 endif()
 
-target_link_libraries(IotSecureTunneling-cpp ${DEP_AWS_LIBS})
```
Comment (fufranci) [inline secure_tunneling/CMakeLists.txt:97]: See line 103. This line seems to be a duplicate.

File: secure_tunneling/source/SecureTunnel.cpp [hunk 2]
```diff
@@ -33,17 +33,21 @@ namespace Aws
...
 
             config.allocator = allocator;
             config.bootstrap = clientBootstrap ? clientBootstrap->GetUnderlyingHandle() : nullptr;
-            config.socket_options = socketOptions ? &socketOptions->GetImpl() : nullptr;
+            config.socket_options = &m_socketOptions.GetImpl();
 
-            config.access_token = aws_byte_cursor_from_c_str(accessToken.c_str());
+            config.access_token = aws_byte_cursor_from_c_str(m_accessToken.c_str());
             config.local_proxy_mode = localProxyMode;
-            config.endpoint_host = aws_byte_cursor_from_c_str(endpointHost.c_str());
```
Comment (fufranci) [inline secure_tunneling/source/SecureTunnel.cpp:46]: The old code stores pointers to parameters which will go out-of-scope when this function returns.

File: secure_tunneling/source/SecureTunnel.cpp [hunk 3]
```diff
@@ -102,37 +106,37 @@ namespace Aws
 
         void SecureTunnel::s_OnConnectionComplete(void *user_data)
         {
-            auto *secureTunnel = static_cast<SecureTunnel *>(user_data);
+            auto *secureTunnel = reinterpret_cast<SecureTunnel *>(user_data);
```
Comment (justinboswell) [inline secure_tunneling/source/SecureTunnel.cpp:109]: Why are these `reinterpret_cast` now? You can `static_cast` just fine from `void*` -> known type

File: secure_tunneling/tests/SecureTunnelTest.cpp [hunk 4]
```diff
@@ -78,7 +79,7 @@ static int before(struct aws_allocator *allocator, void *ctx)
     testContext->secureTunnel = new Aws::Iotsecuretunneling::SecureTunnel(
         allocator,
         nullptr,
-        nullptr,
+        s_socketOptions,
```
Comment (justinboswell) [inline secure_tunneling/tests/SecureTunnelTest.cpp:82]: If you're not going to configure this, you could just pass in `Aws::Crt::Iot::SocketOptions()`

**Review by justinboswell — APPROVED** [review]
Fix-n-ship

File: secure_tunneling/source/SecureTunnel.cpp (hunk 3 above)
Comment (fufranci) [reply secure_tunneling/source/SecureTunnel.cpp:109]: It was suggested by @DavidOgunsAWS a while ago but I never submitted a PR for it. I honestly haven't read about when I should use which.

File: secure_tunneling/tests/SecureTunnelTest.cpp (hunk 4 above)
Comment (fufranci) [reply secure_tunneling/tests/SecureTunnelTest.cpp:82]: Ok. Will do.

File: secure_tunneling/source/SecureTunnel.cpp (hunk 3 above)
Comment (fufranci) [reply secure_tunneling/source/SecureTunnel.cpp]: Undo the change and use `static_cast`.

**Pull request merged**